File: app/src/main/java/io/github/mattsays/rommnative/ui/screen/player/PlayerVisualTheme.kt
package io.github.mattsays.rommnative.ui.screen.player

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.luminance
import io.github.mattsays.rommnative.domain.input.PlayerControlsState
import io.github.mattsays.rommnative.ui.theme.BrandCanvas
import io.github.mattsays.rommnative.ui.theme.BrandPanel
import io.github.mattsays.rommnative.ui.theme.BrandPanelAlt
import io.github.mattsays.rommnative.ui.theme.BrandSeed
import io.github.mattsays.rommnative.ui.theme.BrandText

data class PlayerControlStyle(
    val fillColor: Color,
    val pressedFillColor: Color,
    val borderColor: Color,
    val pressedBorderColor: Color,
    val contentColor: Color,
)

data class PlayerVisualTheme(
    val canvasColor: Color,
    val panelColor: Color,
    val panelAltColor: Color,
    val textColor: Color,
    val accentColor: Color,
    val neutralControlStyle: PlayerControlStyle,
    val faceControlStyles: Map<String, PlayerControlStyle>,
) {
    fun faceStyle(label: String): PlayerControlStyle = faceControlStyles[label.uppercase()] ?: neutralControlStyle
}

fun resolvePlayerVisualTheme(controls: PlayerControlsState?): PlayerVisualTheme {
    val oledBlackMode = controls?.preferences?.oledBlackModeEnabled == true
    val consoleColorsEnabled = controls?.preferences?.consoleColorsEnabled == true
    val familyId = controls?.platformProfile?.familyId
    val canvasColor = if (oledBlackMode) Color.Black else BrandCanvas
    val panelColor = if (oledBlackMode) Color.Black else BrandPanel
    val panelAltColor = if (oledBlackMode) Color(0xFF080808) else BrandPanelAlt
    val accentColor = if (consoleColorsEnabled && familyId != null) {
        consoleNeutralAccent(familyId)
    } else {
        BrandSeed
    }
    val neutralControlStyle = neutralStyle(
        baseColor = panelAltColor,
        accentColor = accentColor,
    )
    val faceStyles = if (consoleColorsEnabled && familyId != null) {
        consoleFaceColors(familyId)
            .mapValues { (_, color) -> coloredStyle(color) }
    } else {
        emptyMap()
    }
    return PlayerVisualTheme(
        canvasColor = canvasColor,
        panelColor = panelColor,
        panelAltColor = panelAltColor,
        textColor = BrandText,
        accentColor = accentColor,
        neutralControlStyle = neutralControlStyle,
        faceControlStyles = faceStyles,
    )
}

private fun neutralStyle(
    baseColor: Color,
    accentColor: Color,
): PlayerControlStyle {
    return PlayerControlStyle(
        fillColor = baseColor.copy(alpha = 0.92f),
        pressedFillColor = accentColor.copy(alpha = 0.38f),
        borderColor = accentColor.copy(alpha = 0.20f),
        pressedBorderColor = accentColor.copy(alpha = 0.64f),
        contentColor = BrandText,
    )
}

private fun coloredStyle(color: Color): PlayerControlStyle {
    val contentColor = if (color.luminance() > 0.55f) Color(0xFF151515) else BrandText
    return PlayerControlStyle(
        fillColor = color.copy(alpha = 0.90f),
        pressedFillColor = color.copy(alpha = 1f),
        borderColor = color.copy(alpha = 0.96f),
        pressedBorderColor = BrandText.copy(alpha = 0.75f),
        contentColor = contentColor,
    )
}

private fun consoleFaceColors(familyId: String): Map<String, Color> {
    return when (familyId) {
        "snes" -> mapOf(
            "Y" to Color(0xFF46B06E),
            "B" to Color(0xFFF1C94B),
            "A" to Color(0xFFDA5A58),
            "X" to Color(0xFF5F7DDE),
        )
        "psx" -> mapOf(
            "SQUARE" to Color(0xFFD06AAF),
            "CROSS" to Color(0xFF4E8EEA),
            "CIRCLE" to Color(0xFFE05C4B),
            "TRIANGLE" to Color(0xFF55B96D),
        )
        "sega16" -> mapOf(
            "A" to Color(0xFF4A78D0),
            "B" to Color(0xFF4FB56B),
            "C" to Color(0xFFF0C14A),
            "X" to Color(0xFFD85E4B),
        )
        "arcade" -> mapOf(
            "1" to Color(0xFF4A78D0),
            "2" to Color(0xFF4FB56B),
            "3" to Color(0xFFF0C14A),
            "4" to Color(0xFFD85E4B),
        )
        "nes" -> mapOf(
            "A" to Color(0xFFD85C56),
            "B" to Color(0xFFB54845),
        )
        "gb" -> mapOf(
            "A" to Color(0xFF8A4FA8),
            "B" to Color(0xFFB45AA2),
        )
        "gba" -> mapOf(
            "A" to Color(0xFF7D70D6),
            "B" to Color(0xFF9887E4),
        )
        "tg16" -> mapOf(
            "I" to Color(0xFF4479CF),
            "II" to Color(0xFFE08C3E),
        )
        "atari" -> mapOf(
            "1" to Color(0xFFE08C3E),
            "2" to Color(0xFFD85E4B),
        )
        else -> emptyMap()
    }
}

private fun consoleNeutralAccent(familyId: String): Color {
    return when (familyId) {
        "snes" -> Color(0xFF7F6BC4)
        "psx" -> Color(0xFF5E82D8)
        "sega16" -> Color(0xFF4A78D0)
        "arcade" -> Color(0xFFD85E4B)
        "nes" -> Color(0xFFB54845)
        "gb" -> Color(0xFF8A4FA8)
        "gba" -> Color(0xFF7D70D6)
        "tg16" -> Color(0xFF4479CF)
        "atari" -> Color(0xFFE08C3E)
        else -> BrandSeed
    }
}

File: ios/Sources/RommioUI/PlayerVisualTheme.swift
import SwiftUI
import RommioFoundation

struct PlayerControlStyle {
    let fillColor: Color
    let pressedFillColor: Color
    let borderColor: Color
    let pressedBorderColor: Color
    let contentColor: Color
}

struct PlayerVisualTheme {
    let canvasColor: Color
    let panelColor: Color
    let panelAltColor: Color
    let textColor: Color
    let accentColor: Color
    let neutralControlStyle: PlayerControlStyle
    let faceControlStyles: [String: PlayerControlStyle]

    func faceStyle(label: String) -> PlayerControlStyle {
        faceControlStyles[label.uppercased()] ?? neutralControlStyle
    }
}

func resolvePlayerVisualTheme(controls: PlayerControlsState?) -> PlayerVisualTheme {
    let oledBlackMode = controls?.preferences.oledBlackModeEnabled == true
    let consoleColorsEnabled = controls?.preferences.consoleColorsEnabled == true
    let familyID = controls?.platformProfile.familyID
    let canvasColor = oledBlackMode ? Color.black : Color(red: 0.05, green: 0.06, blue: 0.08)
    let panelColor = oledBlackMode ? Color.black : Color(red: 0.10, green: 0.11, blue: 0.15)
    let panelAltColor = oledBlackMode ? Color(red: 0.03, green: 0.03, blue: 0.03) : Color(red: 0.14, green: 0.16, blue: 0.20)
    let accentColor = consoleColorsEnabled ? consoleNeutralAccent(familyID: familyID) : Color(red: 0.20, green: 0.52, blue: 0.98)
    let neutral = neutralStyle(baseColor: panelAltColor, accentColor: accentColor)
    let faceStyles = consoleColorsEnabled ? consoleFaceColors(familyID: familyID).mapValues { coloredStyle(color: $0) } : [:]
    return PlayerVisualTheme(
        canvasColor: canvasColor,
        panelColor: panelColor,
        panelAltColor: panelAltColor,
        textColor: .white,
        accentColor: accentColor,
        neutralControlStyle: neutral,
        faceControlStyles: faceStyles
    )
}

private func neutralStyle(baseColor: Color, accentColor: Color) -> PlayerControlStyle {
    PlayerControlStyle(
        fillColor: baseColor.opacity(0.92),
        pressedFillColor: accentColor.opacity(0.38),
        borderColor: accentColor.opacity(0.20),
        pressedBorderColor: accentColor.opacity(0.64),
        contentColor: .white
    )
}

private func coloredStyle(color: Color) -> PlayerControlStyle {
    PlayerControlStyle(
        fillColor: color.opacity(0.90),
        pressedFillColor: color.opacity(1),
        borderColor: color.opacity(0.96),
        pressedBorderColor: .white.opacity(0.75),
        contentColor: color.brightness > 0.55 ? Color(red: 0.08, green: 0.08, blue: 0.08) : .white
    )
}

private func consoleFaceColors(familyID: String?) -> [String: Color] {
    switch familyID {
    case "snes":
        ["Y": Color(hex: 0x46B06E), "B": Color(hex: 0xF1C94B), "A": Color(hex: 0xDA5A58), "X": Color(hex: 0x5F7DDE)]
    case "psx":
        ["SQUARE": Color(hex: 0xD06AAF), "CROSS": Color(hex: 0x4E8EEA), "CIRCLE": Color(hex: 0xE05C4B), "TRIANGLE": Color(hex: 0x55B96D)]
    case "sega16":
        ["A": Color(hex: 0x4A78D0), "B": Color(hex: 0x4FB56B), "C": Color(hex: 0xF0C14A), "X": Color(hex: 0xD85E4B)]
    case "arcade":
        ["1": Color(hex: 0x4A78D0), "2": Color(hex: 0x4FB56B), "3": Color(hex: 0xF0C14A), "4": Color(hex: 0xD85E4B)]
    case "nes":
        ["A": Color(hex: 0xD85C56), "B": Color(hex: 0xB54845)]
    case "gb":
        ["A": Color(hex: 0x8A4FA8), "B": Color(hex: 0xB45AA2)]
    case "gba":
        ["A": Color(hex: 0x7D70D6), "B": Color(hex: 0x9887E4)]
    case "tg16":
        ["I": Color(hex: 0x4479CF), "II": Color(hex: 0xE08C3E)]
    case "atari":
        ["1": Color(hex: 0xE08C3E), "2": Color(hex: 0xD85E4B)]
    default:
        [:]
    }
}

private func consoleNeutralAccent(familyID: String?) -> Color {
    switch familyID {
    case "snes":
        Color(hex: 0x7F6BC4)
    case "psx":
        Color(hex: 0x5E82D8)
    case "sega16":
        Color(hex: 0x4A78D0)
    case "arcade":
        Color(hex: 0xD85E4B)
    case "nes":
        Color(hex: 0xB54845)
    case "gb":
        Color(hex: 0x8A4FA8)
    case "gba":
        Color(hex: 0x7D70D6)
    case "tg16":
        Color(hex: 0x4479CF)
    case "atari":
        Color(hex: 0xE08C3E)
    default:
        Color(red: 0.20, green: 0.52, blue: 0.98)
    }
}

private extension Color {
    init(hex: UInt32, alpha: Double = 1) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: alpha
        )
    }

    var brightness: Double {
        #if canImport(UIKit)
        let uiColor = UIColor(self)
        var hue: CGFloat = 0
        var saturation: CGFloat = 0
        var brightness: CGFloat = 0
        var alpha: CGFloat = 0
        uiColor.getHue(&hue, saturation: &saturation, brightness: &brightness, alpha: &alpha)
        return brightness
        #else
        return 0.5
        #endif
    }
}
